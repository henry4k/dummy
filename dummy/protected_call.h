#ifndef __DUMMY_PROTECTED_CALL_H__
#define __DUMMY_PROTECTED_CALL_H__

#ifdef __cplusplus
extern "C"
{
#endif

typedef void (*dummyProtectableFunction)();

typedef void (*dummyCleanupFunction)( void* data );

enum
{
    DUMMY_PROTECTED_CALL_SUCEEDED = 0,
    DUMMY_PROTECTED_CALL_GENERIC_ERROR = 1
};

/**
 * Runs the given function and catches errors.
 *
 * Protected calls can be nested.
 *
 * @param abortReason
 * If the protected call is aborted with a reason,
 * the string pointer is passed here.
 *
 * @return
 * Code that classifies the error.
 * If the function ras successfully it returns DUMMY_PROTECTED_CALL_SUCEEDED.
 * Custom error codes may be passed using dummyAbortProtectedCall().
 */
int dummyProtectedCall( dummyProtectableFunction fn, const char** abortReason );

/**
 * Adds a cleanup function to the current protected call.
 *
 * @param data
 * Data pointer that is passed to the cleanup function.
 */
void dummyAddCleanup( dummyCleanupFunction fn, void* data );

/**
 * Aborts the current protected call.
 *
 * @param reason
 * May be `NULL`.
 *
 * @return
 * Doesn't return.
 */
void dummyAbortProtectedCall( int errorCode, const char* reason );

#ifdef __cplusplus
}
#endif

#endif
