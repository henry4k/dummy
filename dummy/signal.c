#include <assert.h>

#include "core.h" // DUMMY_MAX_SANDBOX_DEPTH
#include "signal.h"


static dummySignalHandler signalHandlerStack[DUMMY_MAX_SANDBOX_DEPTH];
static int signalHandlerStackSize = 0;


static void setSignalHandler( dummySignalHandler fn )
{
    signal(SIGABRT, fn);
    signal(SIGFPE, fn);
    signal(SIGILL, fn);
    signal(SIGSEGV, fn);
}

void dummyPushSignalHandler( dummySignalHandler fn )
{
    assert(signalHandlerStackSize < DUMMY_MAX_SANDBOX_DEPTH);
    signalHandlerStackSize++;
    signalHandlerStack[signalHandlerStackSize-1] = fn;

    setSignalHandler(fn);
}

void dummyPopSignalHandler()
{
    assert(signalHandlerStackSize > 0);
    signalHandlerStackSize--;
    if(signalHandlerStackSize > 0)
        setSignalHandler(signalHandlerStack[signalHandlerStackSize-1]);
    else
        setSignalHandler(SIG_DFL);
}

const char* dummySignalToAbortReason( int signal )
{
    switch(signal)
    {
        case SIGABRT:
            return "SIGABRT: Abnormal termination";

        case SIGILL:
            return "SIGILL: Illegal instruction";

        case SIGSEGV:
            return "SIGSEGV: Segmentation violation";

        case SIGFPE:
            //const int exception = fetestexcept(FE_ALL_EXCEPT);
            //if(exception & FE_DIVBYZERO) puts("FE_DIVBYZERO");
            //if(exception & FE_INEXACT)   puts("FE_INEXACT");
            //if(exception & FE_INVALID)   puts("FE_INVALID");
            //if(exception & FE_OVERFLOW)  puts("FE_OVERFLOW");
            //if(exception & FE_UNDERFLOW) puts("FE_UNDERFLOW");
            return "SIGFPE: Floating point exception";

        default:
            return "An unknown signal has been raised.";
    }
}
