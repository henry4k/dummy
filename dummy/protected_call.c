#include <stddef.h> // NULL
#include <signal.h>
#include <setjmp.h>
#include <string.h> // memset, strncpy
#include <fenv.h> // floating point exceptions
#include <assert.h>

#include "protected_call.h"


enum
{
    DUMMY_MAX_ENVIRONMENTS = 16,
    DUMMY_MAX_PROTECTED_CALL_REASON_LENGTH = 512
};

typedef struct
{
    jmp_buf jumpBuffer;
    int errorCode;
    char abortReason[DUMMY_MAX_PROTECTED_CALL_REASON_LENGTH];
} dummyEnvironment;

typedef void (*dummySignalFunction)(int);


dummyEnvironment dummyEnvironmentStack[DUMMY_MAX_ENVIRONMENTS];
int dummyEnvironmentStackSize = 0;


void dummySignalHandler( int signal );
void dummySetAbortInformation( int errorCode, const char* reason );

void dummySetSignals( dummySignalFunction fn )
{
    signal(SIGABRT, fn);
    signal(SIGFPE, fn);
    signal(SIGILL, fn);
    signal(SIGSEGV, fn);
}

dummyEnvironment* dummyGetCurrentEnvironment()
{
    if(dummyEnvironmentStackSize > 0)
        return &dummyEnvironmentStack[dummyEnvironmentStackSize-1];
    else
        return NULL;
}

dummyEnvironment* dummyPushEnvironment()
{
    if(dummyEnvironmentStackSize == 0)
        dummySetSignals(dummySignalHandler);

    dummyEnvironmentStackSize++;
    assert(dummyEnvironmentStackSize <= DUMMY_MAX_ENVIRONMENTS);

    dummyEnvironment* environment = dummyGetCurrentEnvironment();
    memset(environment, 0, sizeof(dummyEnvironment));
    environment->errorCode = DUMMY_PROTECTED_CALL_SUCEEDED;
    return environment;
}

dummyEnvironment* dummyPopEnvironment()
{
    assert(dummyEnvironmentStackSize > 0);
    dummyEnvironment* environment = dummyGetCurrentEnvironment();
    dummyEnvironmentStackSize--;
    if(dummyEnvironmentStackSize == 0)
        dummySetSignals(SIG_DFL);
    return environment;
}

const char* dummySignalToAbortReason( int signal )
{
    switch(signal)
    {
        case SIGABRT:
            return "SIGABRT: Abnormal termination";

        case SIGILL:
            return "SIGILL: Illegal instruction";

        case SIGSEGV:
            return "SIGSEGV: Segmentation violation";

        case SIGFPE:
            //const int exception = fetestexcept(FE_ALL_EXCEPT);
            //if(exception & FE_DIVBYZERO) puts("FE_DIVBYZERO");
            //if(exception & FE_INEXACT)   puts("FE_INEXACT");
            //if(exception & FE_INVALID)   puts("FE_INVALID");
            //if(exception & FE_OVERFLOW)  puts("FE_OVERFLOW");
            //if(exception & FE_UNDERFLOW) puts("FE_UNDERFLOW");
            return "SIGFPE: Floating point exception";

        default:
            return NULL;
    }
}

void dummySignalHandler( int signal )
{
    dummyEnvironment* environment = dummyGetCurrentEnvironment();
    if(environment)
    {
        // If dummySetAbortInformation was not called before:
        if(environment->errorCode == DUMMY_PROTECTED_CALL_SUCEEDED)
        {
            dummySetAbortInformation(
                DUMMY_PROTECTED_CALL_GENERIC_ERROR,
                dummySignalToAbortReason(signal));
        }

        longjmp(environment->jumpBuffer, 1);
    }
}

int dummyProtectedCall( dummyProtectableFunction fn, const char** abortReason )
{
    dummyEnvironment* environment = dummyPushEnvironment();
    const int jumpResult = setjmp(environment->jumpBuffer);
    if(jumpResult == 0)
    {
        fn();
    }
    else
    {
        // A signal has been catched: reset signals
        dummySetSignals(dummySignalHandler);
    }
    dummyPopEnvironment();

    if(abortReason)
    {
        if(environment->abortReason[0] == '\0')
            *abortReason = NULL;
        else
            *abortReason = environment->abortReason;
    }
    return environment->errorCode;
}

void dummySetAbortInformation( int errorCode, const char* reason )
{
    dummyEnvironment* environment = dummyGetCurrentEnvironment();
    assert(environment);

    environment->errorCode = errorCode;
    if(reason)
        strncpy(environment->abortReason, reason, DUMMY_MAX_PROTECTED_CALL_REASON_LENGTH);
}

void dummyAbortProtectedCall( int errorCode, const char* reason )
{
    dummySetAbortInformation(errorCode, reason);
    raise(SIGABRT);
}
