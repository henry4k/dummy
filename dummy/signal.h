/**
 * This file just contains tools for signal and exception runner.
 * You don't want to use it directly.
 */
#ifndef __DUMMY_SIGNAL_H__
#define __DUMMY_SIGNAL_H__

#include <signal.h>


typedef void (*dummySignalFunction)(int);

void dummySetSignals( dummySignalFunction fn );
const char* dummySignalToAbortReason( int signal );

#endif
