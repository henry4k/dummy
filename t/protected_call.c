#include <stdio.h>
#include <signal.h>

#include "../dummy/protected_call.h"


void pcall_2()
{
    raise(SIGABRT);
}

void pcall_1()
{
    const char* abortReason = "bbb";
    const int errorCode = dummyProtectedCall(pcall_2, &abortReason);
    if(errorCode != DUMMY_PROTECTED_CALL_SUCEEDED)
        dummyAbortProtectedCall(errorCode, abortReason);
}

int main()
{
    const char* abortReason = "aaa";
    const int errorCode = dummyProtectedCall(pcall_1, &abortReason);
    printf("error code = %d, abort reason = '%s'\n",
        errorCode,
        abortReason);
    return 0;
}
