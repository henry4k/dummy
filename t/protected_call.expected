error code = 1, abort reason = 'SIGABRT: Abnormal termination'
